packages/theme/modules/catalog/product/types.ts

```diff
@@ -0,0 +1,39 @@
+import {
+  BundleProduct as BaseBundleProduct,
+  ConfigurableProduct as BaseConfigurableProduct,
+  DownloadableProduct as BaseDownloadableProduct,
+  GroupedProduct as BaseGroupedProduct, ProductAttributeFilterInput, ProductAttributeSortInput,
+  ProductInterface,
+  VirtualProduct as BaseVirtualProduct,
+} from '~/modules/GraphQL/types';
+
+/**
+ * There is no __typename in GraphQL definitions but type_id is marked as a deprecated
+ */
+export interface Product extends ProductInterface, ConfigurableProduct, Omit<BundleProduct, 'items'>, Omit<GroupedProduct, 'items'>, Omit<DownloadableProduct, 'items'>, Omit<VirtualProduct, 'items'> { __typename: string }
+export interface BundleProduct extends BaseBundleProduct { __typename: string }
+export interface ConfigurableProduct extends BaseConfigurableProduct { __typename: string }
+export interface DownloadableProduct extends BaseDownloadableProduct { __typename: string }
+export interface ConfigurableProduct extends BaseConfigurableProduct { __typename: string }
+export interface GroupedProduct extends BaseGroupedProduct { __typename: string }
+export interface VirtualProduct extends BaseVirtualProduct { __typename: string }
```

Comment:
```suggestion
export interface VirtualProduct extends BaseVirtualProduct { __typename: string }
```

You could avoid creating all those interfaces by just creating a generic type
```ts
type WithTypename<TProduct> = TProduct & { __typename: string };
```
and then in code using them like
```
product: Object as PropType<WithTypename<BaseConfigurableProduct>>
```

If there is some "super base product interface shared between all products" (knowing graphql codegen probably not), you could also do `type WithTypename<TProduct extends SuperBaseProduct>` aswell